In [1]:
import fastf1

In [2]:
import pandas as pd

In [3]:
import sys
sys.path.append('../src')

from data_loader import processar_quali

In [4]:
processar_quali(2025, 'Bahrain')

core           INFO 	Loading data for Bahrain Grand Prix - Qualifying [v3.8.3]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
req            INFO 	Using cached data for car_data
req            INFO 	Using cached data for position_data
req            INFO 	Using cached data for weather_data
req            INFO 	Using cached data for race_control_messages
core           INFO 	Finished loading data for 20 drivers: ['81', '63', '16', '12', '10', '4', '1', '55', '44', '22', '7', '6', '14', '31', '23', '27', '30', '5', '18', '87']


,Piloto,Equipe,S1,S2,S3,Tempo de Volta,Position,Ano,GP
0,PIA,McLaren,28.784,38.574,22.483,89.841,1,2025,Bahrain Grand Prix
1,RUS,Mercedes,28.771,38.710,22.528,90.009,2,2025,Bahrain Grand Prix
2,LEC,Ferrari,28.926,38.714,22.535,90.175,3,2025,Bahrain Grand Prix
3,ANT,Mercedes,28.836,38.786,22.591,90.213,4,2025,Bahrain Grand Prix
4,GAS,Alpine,28.883,38.694,22.639,90.216,5,2025,Bahrain Grand Prix
5,NOR,McLaren,29.007,38.678,22.582,90.267,6,2025,Bahrain Grand Prix
6,VER,Red Bull Racing,28.880,38.855,22.688,90.423,7,2025,Bahrain Grand Prix
7,SAI,Williams,28.894,39.063,22.723,90.680,8,2025,Bahrain Grand Prix
8,HAM,Ferrari,28.955,39.058,22.759,90.772,9,2025,Bahrain Grand Prix
9,TSU,Red Bull Racing,29.067,39.177,22.984,91.228,10,2025,Bahrain Grand Prix


## --------------------------------------------------------------------------------------- ##

In [5]:
calendario = fastf1.get_event_schedule(2025)
calendario[['RoundNumber', 'EventName', 'Country', 'EventDate']]

,RoundNumber,EventName,Country,EventDate
0,0,Pre-Season Testing,Bahrain,2025-02-28
1,1,Australian Grand Prix,Australia,2025-03-16
2,2,Chinese Grand Prix,China,2025-03-23
3,3,Japanese Grand Prix,Japan,2025-04-06
4,4,Bahrain Grand Prix,Bahrain,2025-04-13
5,5,Saudi Arabian Grand Prix,Saudi Arabia,2025-04-20
6,6,Miami Grand Prix,United States,2025-05-04
7,7,Emilia Romagna Grand Prix,Italy,2025-05-18
8,8,Monaco Grand Prix,Monaco,2025-05-25
9,9,Spanish Grand Prix,Spain,2025-06-01


In [6]:
corridas = calendario[calendario['RoundNumber'] > 0]
corridas[['RoundNumber', 'EventName']]

,RoundNumber,EventName
1,1,Australian Grand Prix
2,2,Chinese Grand Prix
3,3,Japanese Grand Prix
4,4,Bahrain Grand Prix
5,5,Saudi Arabian Grand Prix
6,6,Miami Grand Prix
7,7,Emilia Romagna Grand Prix
8,8,Monaco Grand Prix
9,9,Spanish Grand Prix
10,10,Canadian Grand Prix


In [7]:
tabelas = []
falhas = []

for rodada in corridas['RoundNumber']:
    try:
         tabela = processar_quali(2025, rodada)
         tabelas.append(tabela)
    except Exception as e:
        print(f"Falha na rodada {rodada}: {e}")
        falhas.append(rodada)

len(tabelas), falhas



core           INFO 	Loading data for Australian Grand Prix - Qualifying [v3.8.3]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
core        WARNING 	Fixed incorrect tyre stint information for driver '4'
req            INFO 	Using cached data for car_data
req            INFO 	Using cached data for position_data
req            INFO 	Using cached data for weather_data
req            INFO 	Using cached data for race_control_messages
core           INFO 	Finished loading data for 20 drivers: ['4', '81', '1', '63', '22', '23', '16', '44', '10', '55', '6', '14', '18', '7', '5', '12', '27', '30', '31', '87']
core           INFO 	Loading dat

(24, [])

In [8]:
processar_quali(2025, 1)

core           INFO 	Loading data for Australian Grand Prix - Qualifying [v3.8.3]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
core        WARNING 	Fixed incorrect tyre stint information for driver '4'
req            INFO 	Using cached data for car_data
req            INFO 	Using cached data for position_data
req            INFO 	Using cached data for weather_data
req            INFO 	Using cached data for race_control_messages
core           INFO 	Finished loading data for 20 drivers: ['4', '81', '1', '63', '22', '23', '16', '44', '10', '55', '6', '14', '18', '7', '5', '12', '27', '30', '31', '87']


,Piloto,Equipe,S1,S2,S3,Tempo de Volta,Position,Ano,GP
0,NOR,McLaren,25.961,16.997,32.138,75.096,1,2025,Australian Grand Prix
1,PIA,McLaren,26.082,16.952,32.146,75.180,2,2025,Australian Grand Prix
2,VER,Red Bull Racing,26.066,16.915,32.500,75.481,3,2025,Australian Grand Prix
3,RUS,Mercedes,26.163,16.951,32.432,75.546,4,2025,Australian Grand Prix
4,TSU,Racing Bulls,26.190,17.005,32.475,75.670,5,2025,Australian Grand Prix
5,ALB,Williams,26.215,16.978,32.544,75.737,6,2025,Australian Grand Prix
6,LEC,Ferrari,26.194,17.025,32.536,75.755,7,2025,Australian Grand Prix
7,HAM,Ferrari,26.150,17.223,32.546,75.919,8,2025,Australian Grand Prix
8,GAS,Alpine,26.293,17.087,32.600,75.980,9,2025,Australian Grand Prix
9,SAI,Williams,26.269,17.078,32.584,75.931,10,2025,Australian Grand Prix


In [9]:
df_temporada = pd.concat(tabelas, ignore_index=True)
df_temporada.shape

(475, 9)

In [10]:
df_temporada['GP'].nunique()

24

In [11]:
df_temporada.to_csv('../data/processed/quali_2025.csv', index=False)

## Coleta de dados — temporada 2025

**Objetivo:** a função `processar_quali(ano, gp)` recebe o ano e o GP (nome ou número da rodada) e devolve uma tabela com uma linha por piloto, contendo equipe, tempos de setor (S1, S2, S3), tempo de volta e posição de grid.

**Como funciona:**
- Carrega a sessão de quali daquele GP e pega a melhor volta cronometrada de cada piloto
- Junta essa tabela com `session.results` pela abreviação do piloto (nunca pela ordem das linhas — a ordem pode não significar nada)
- Converte os tempos, que vêm como `Timedelta`, para segundos com `.total_seconds()`

**Bugs encontrados e corrigidos:**
1. **Piloto sem volta rápida:** `pick_fastest()` devolve `None` quando o piloto não tem nenhuma volta cronometrada válida na sessão (ex: Bearman bateu no treino livre da Austrália 2025 e não rodou no quali). Sem checar isso, a linha seguinte (`volta['Driver']`) quebra com `'NoneType' object is not subscriptable`. A correção pula esse piloto com `if volta is None: continue`.

**Loop pela temporada:** percorri as 24 corridas de 2025 (filtrando a Rodada 0, que é a pré-temporada e não tem quali) com `try/except`, pra garantir que uma falha numa corrida não derrubasse o processamento das outras.

**Resultado:** 475 linhas de um teórico máximo de 480 (24 corridas × 20 pilotos) — a diferença de 5 vem de pilotos sem volta cronometrada em alguma sessão ao longo do ano.

**Onde ficou salvo:** `data/processed/quali_2025.csv`